### ISSUE TITLE
Why not adding nonce/salt to file header?

### ISSUE THREAD

**Issue by fmpwizard** [minio/sio#5]
Maybe I misunderstood something from the announcement but, if I encrypt 1000 files, I need to create 1000 nonces and then I somehow need to keep track of each nonce so I can decrypt each file?

Keeping track of each nonce seems the kind of requirement that user may either end up reusing the same nonce or may misplace it, resulting in not being able to decrypt files any more.

If there a reason why your format doesn't include adding the nonce to the file itself? similar to how using bcrypt, the salt is prefixed to the ciphertext?

**Comment by cstockton** [comment]
As I mentioned [on reddit](https://www.reddit.com/r/golang/comments/6yoyar/go_implementation_of_data_at_rest_encryption/dmp8ss4/) I'm also interested in a more formal response to this decision.

**Comment by aead** [comment]
In general yes, if you encrypt 1000 files with the same (master) key DARE requires 1000 unique encryption keys to be tamper proof. There are many approaches to derive unique keys from a (master) key. Your approach is the simplest but most expensive on - you have to save 1000 nonces each nonce linked to the correct file.
A different approach would be to e.g. derive the encryption key of the file from it's name. 
`key = KDF(masterkey, filename)` With this solution you don't need to save any nonce at all. The downside is that you cannot rename an encrypted file and that whenever you replace a file (e.g. overwrite it) using the same master key you lose the tamper-proof property again. The key management is more or less specific and depends highly on the requirements. 

About including a salt / kdf-nonce into the package header itself:
TL;DR **It does not improve security / solve the key-management issue.**

Let's assume we have a salt in each package header:
`header := version || cipher || payload_len || seq_number || nonce || salt`
We choose a reasonable size for the salt e.g. 24 bytes. (Let's not worry about why there is still a nonce for now). 

Now we assume we have two different files *f_1* and *f_2* (to keep things simple 100 bytes long).
We encrypt each file with the master key (and AES-GCM): *key_m*. The encryption key for every package is derived like this: `key := KDF(key_m, salt_i)`. The result are two packages *p_1* and *p_2*:
```
p_1 = Encrypt(f_1, key_m)
p_2 = Encrypt(f_2, key_m)
```
The headers will look like this:
```
h_1 := 0x10 || 0 || 100 || 0 || nonce_1 || salt_1
h_2 :=  0x10 || 0 || 100 || 0 || nonce_2 || salt_2 
``` 
Now assume somebody swaps *p_1* and *p_2* so: `p_1 := p_2 and p_2 := p_1`. The problem: We won't detect that both files have changed. Imagine both files are executables - you way don't want to execute *f_2* while you think you run *f_1*... 

Further even a salt would not prevent exchanging packages with the same sequence number:
Let's assume we have a two sequences of packages encrypted with the same master key:
```
package_a_0 || package_a_1 || package_a_2
package_b_0 || package_b_1 || package_b_2
```
Now somebody can exchange e.g. *package_a_1* and *package_b_1* - we would not detect it because both packages (a_1 and b_1) are valid, they "just" don't belong to the sequence. 

We could fix this by fixing the salt per data stream - like: The salt **MUST** be the same for all packages within a sequence of packages. Unfortunately this is would exclude appending.
Appending to an existing sequence is not possible without reading and verifying the whole encrypted data stream first -> Only plaintext append would be possible.

Furthermore there is a serious problem with random access reads. Imagine we only want to decrypt one package of a sequence of packages. How do we verify that this package was not replaced by a different package of another sequence encrypted with the same master key? We can't verify that the salt of package is "correct". We could read two packages instead of one and check that both salt values are the same but how do we ensure that the attacker didn't replace both packages. Same issue for 3, 4, 5 ... packages.
This goes on until we read the whole data stream and verified that all salt values are equal. But now we do not access data randomly - we always read the whole data stream.

Conclusion:
A salt would not improve security. As far as I can see DARE is tamper-proof when the master key is unique and not tamper-proof when the master key is reused - the salt doesn't change this.

**Comment by fmpwizard** [comment]
Thanks for the long reply, but I think my original question wasn't clear. I wasn't implying that adding the nonce to the encrypted file was going to improve security. I was under the impression that one of the goals of this library was to be used by most developers, but as I saw on your reddit reply to another question, this library is only meant for those who already know about cryptography.
It would be good to add such a sense to the README, so people know what they are getting into.

To clarify, my point is that asking regular non-crypto developers to keep a secret master key safe is acceptable, ok and expected, but asking them to generate a unique nonce per file and keep track of each nonce is asking too much.

Thank you.

**Comment by aead** [comment]
Thanks for the hint. I will add such a statement to the README. DARE is much more a primitive than a of-the-shelf solution because of the mentioned challenges. 
Just to share some information about this problem. Next week(s) we will add a CLI tool (similar to gpg, mcrypt etc.) which solves this problem. The user won't deal with nonces at all while still keeping the data confidential and tamper-proof. 
 
And yes I agree - non-crypto developers should not deal with the "unique-key / nonce -problem".

**Comment by aead** [comment]
Closing this now - If there are further questions - reopen or open new issue.

**Comment by hemancuso** [comment]
This remains a bit unclear to me - specifically as to whether or not a public nonce [sitting alongside an encrypted stream] weakens security as long as the key remains private.

Can we pick a random nonce and use that for both the filename and the encryption nonce without weakening security? I'm also trying to assemble an of DARE encoded objects into a single object [S3 multipart upload] and it's not clear to me if this encryption model would let me wrap each object with a header that specifies the length and the nonce followed by the SIO encrypted payload. Thanks for any clarification, but I'd prefer not to design-in a mistake!

**Comment by aead** [comment]
> This remains a bit unclear to me - specifically as to whether or not a public nonce [sitting alongside an encrypted stream] weakens security as long as the key remains private.

A public nonce value does not weaken the security since the corresponding AEAD schemes (AES-GCM / ChaCha20Poly1305) are considered secure against a nonce-respecting adversary in the IND-CPA model. So a nonce value can be public information - as long as it is not repeated for the same key it is fine.

> Can we pick a random nonce and use that for both the filename and the encryption nonce without weakening security?

What do you mean by using the nonce for the filename? Create an DAREv2 encrypted file and re-using the initial nonce value for the filename?

> I'm also trying to assemble a stream of DARE encoded objects into a single object [S3 multipart upload] and it's not clear to me if this encryption model would let me wrap each object with a header that specifies the length and the nonce followed by the SIO encrypted payload. 

Let me understand the scenario. Are you trying to create a single DAREv2 encrypted object as an S3 client? So you have uploaded e.g. 3 encrypted objects and want to compose them into one single object? Or are you trying to implement S3-like multipart (as a server) such that you get an DAREv2 encrypted object?
